src/task4feedback/simulator/rl/models/simple.py

```diff
@@ -0,0 +1,295 @@
+            if self.is_training_mode():
+                self.update_pi(state, state_tuple, oracle)
+        return action
+
+    def update_pi(self, state: torch.tensor, state_tuple: Tuple[float], oracle):
+        """ Update target policy, pi.
+        """
+
+        # Buffer probabilities from network
+        actions, v = self.network(NetworkInput(state, False, None, None))
+        actions = F.softmax(actions, dim=0)
+        # print("oracle:", oracle, " softmax dnn:", actions)
+        # Calculate loss
+        self.pi[state_tuple] = (1 - self.ld) * actions + self.ld * oracle.to(self.device)
+
+    def add_reward(self, reward):
+        """
+        Add a reward to the list.
+        """
+        self.accumulated_reward += reward.item()
+
+    def optimize_model(self, reward: float):
+        """
+        Optimize a model.
+        """
+        print("Model optimization starts..")
+        prob_loss = 0
+        value_loss = 0
+        num_computes = 0
+        entropy_sum = 0
+        plist = []
+        vlist = []
+        pilist = []
+        zlist = []
+        for log in self.logs:
+            state, next_state = log
+            state_tuple = tuple(state.tolist())
+            p, v = self.network(NetworkInput(state, False, None, None))
+            p = F.softmax(p, dim=0)
+            pis = self.pi[state_tuple]
+            # print("p:", p, " pis:", pis)
+            plist.append(p)
+            vlist.append(v)
+            pilist.append(pis)
+            zlist.append(torch.tensor([reward], dtype=torch.float))
+
+        concat_p = torch.cat(
+            [p.unsqueeze(0) for p in plist])
+        # concat_p = torch.FloatTensor(plist)
+        concat_v = torch.cat(vlist).to(self.device)
+        # Pi is not updated through this
+        concat_pi = torch.cat(
+            [pi.unsqueeze(0) for pi in pilist]).to(self.device).detach()
+        # concat_pi = torch.FloatTensor(pilist)
+        concat_z = torch.cat(zlist).to(self.device)
+        # print("plist:", plist, " pilist:", pilist)
+        """
+        print("concat p:", concat_p)
+        print("concat v:", concat_v)
+        print("concat pi:", concat_pi)
+        print("concat z:", concat_z)
+
+        # print("usq concat p:", concat_p.unsqueeze(-1))
+        print("usq concat v:", concat_v.unsqueeze(-1))
+        # print("usq concat pi:", concat_pi.unsqueeze(-1))
+        print("usq concat z:", concat_z.unsqueeze(-1))
+        with torch.no_grad():
+            print("crossentropy:", F.cross_entropy(concat_p, concat_pi))
+            print("crossentropy mean:", F.cross_entropy(concat_p, concat_pi).mean())
+            # print("usq crossentropy:", F.cross_entropy(concat_p.unsqueeze(-1), concat_pi.unsqueeze(-1)))
+            print("usq mse loss:", F.mse_loss(concat_v.unsqueeze(-1), concat_z.unsqueeze(-1)))
+            print("mse loss:", F.mse_loss(concat_v, concat_z))
+        """
+
+        loss = -F.cross_entropy(concat_p, concat_pi, reduction='sum') + \
```

Comment:
No regularization?

Reply:
If I understood correctly, if I set an weight decay parameter to the optimizer, L2 penalty will be added to the weight. I will add it.